staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types.go

```diff
@@ -174,6 +174,17 @@ type CustomResourceDefinitionVersion struct {
 	// storage indicates this version should be used when persisting custom resources to storage.
 	// There must be exactly one version with storage=true.
 	Storage bool `json:"storage" protobuf:"varint,3,opt,name=storage"`
+	// deprecated indicates this version of the custom resource API is deprecated.
+	// When set to true, API requests to this version receive a warning header in the server response.
+	// Defaults to false.
+	// +optional
+	Deprecated bool `json:"deprecated,omitempty" protobuf:"varint,7,opt,name=deprecated"`
+	// deprecationWarning overrides the default warning returned to API clients.
+	// May only be set when `deprecated` is true.
+	// The default warning indicates this version is deprecated and recommends use
+	// of the newest served version of equal or greater stability, if one exists.
+	// +optional
+	DeprecationWarning string `json:"deprecationWarning,omitempty" protobuf:"bytes,8,opt,name=deprecationWarning"`
```

Comment:
sigh. I guess I could see it for the deprecationWarning field, since explicitly setting an empty string would ignore your explicit value, which isn't great, I guess.

for `deprecation`, it's a bool, so there will only ever be two values, and we will never default deprecated to `true`, so I think a non-pointer is fine

Reply:
I agree for `deprecation` it's fine. For `deprecationWarning`, I don't see why a user would explicitly want an empty warning message.

The question was more about api conventions instead of real use cases. I don't feel strongly about it

/lgtm

Reply:
switched deprecationWarning to a pointer and updated validation to disallow explicit empty string values